org.adempiere.base/src/org/compiere/model/MCostDetail.java

```diff
@@ -1411,4 +1733,71 @@ else if (getM_MatchInv_ID() > 0)
 		return cost.save();
 	}	//	process
 	
+	/**
+	 * Get Cost History Record
+	 * @param as accounting schema
+	 * @return X_M_CostHistory or null
+	 */
+	public X_M_CostHistory getCostHistory(MAcctSchema as)
+	{
+		int AD_Org_ID = getAD_Org_ID();
+		int M_AttributeSetInstance_ID = getM_AttributeSetInstance_ID();
+		String costingMethod = null;
+
+		MProduct product = MProduct.get(getCtx(), getM_Product_ID(), get_TrxName());
+		String CostingLevel = product.getCostingLevel(as);
+		if (MAcctSchema.COSTINGLEVEL_Client.equals(CostingLevel))
+		{
+			AD_Org_ID = 0;
+			M_AttributeSetInstance_ID = 0;
+		}
+		else if (MAcctSchema.COSTINGLEVEL_Organization.equals(CostingLevel))
+			M_AttributeSetInstance_ID = 0;
+		else if (MAcctSchema.COSTINGLEVEL_BatchLot.equals(CostingLevel))
+			AD_Org_ID = 0;
+		//	Costing Method
+		if (costingMethod == null)
+		{
+			costingMethod = product.getCostingMethod(as);
+			if (costingMethod == null)
+			{
+				throw new IllegalArgumentException("No Costing Method");
+			}
+		}
+		
+		if (MCostElement.COSTINGMETHOD_StandardCosting.equals(costingMethod))
+			return null;
+		
+		StringBuilder sql0 = new StringBuilder();
+		sql0.append("SELECT c.M_CostHistory_ID ");
+		sql0.append("FROM M_CostMovement_v c ");
+		sql0.append("JOIN M_CostDetail cd ON (cd.M_CostDetail_ID = c.M_CostDetail_ID AND cd.Processed=?) ");
+		sql0.append("LEFT OUTER JOIN M_CostElement ce ON (c.M_CostElement_ID=ce.M_CostElement_ID) ");
+		sql0.append("WHERE c.AD_Client_ID=? AND c.AD_Org_ID=? ");
+		sql0.append(" AND c.M_Product_ID=? ");
+		sql0.append(" AND (c.M_AttributeSetInstance_ID=? OR c.M_AttributeSetInstance_ID=0) ");
+		sql0.append(" AND c.M_CostType_ID=? AND c.C_AcctSchema_ID=? ");
+		sql0.append(" AND (ce.CostingMethod IS NULL OR ce.CostingMethod=?) ");
+		sql0.append(" AND c.M_CostDetail_ID=? ");
+		sql0.append("ORDER BY c.DateAcct DESC, c.M_CostDetail_ID DESC, c.M_CostHistory_ID DESC");
```

Comment:
this could be a performance hotspot, please analyse whether there's more efficient solution for this.